src/util/mergeUpdates.ts

```diff
@@ -16,26 +16,79 @@ const mergeUpdates = <T>(
   // assume an optional pending property
   type MaybePending = T & { pending?: boolean }
 
-  const mergeResult = { ...mergeInto }
-
-  Object.entries(mergee).forEach(([key, value]) => {
-    if (value) {
-      // ignore pending objects that would overwrite non-pending objects by default
-      if (
-        overwritePending ||
-        !(value as MaybePending).pending ||
-        !mergeInto[key] ||
-        (mergeInto[key] as MaybePending).pending
-      ) {
-        mergeResult[key] = value
+  /** Determine if we should allow overwrite based on pending logic. */
+  const shouldOverwrite = (currentPending: boolean | undefined, incomingPending: boolean | undefined) => {
+    const hasIncomingPending = incomingPending !== undefined
+    // Scenarios:
+    // 1. overwritePending is true: always allow overwrite (original behaviour).
+    // 2. currentPending === true: allow overwrite so that pending can be cleared.
+    // 3. currentPending === false && incomingPending === true: block overwrite to keep non-pending state.
+    // 4. currentPending === false && incomingPending === false/undefined: allow overwrite (other fields may update).
+    return overwritePending || currentPending !== false || (hasIncomingPending && incomingPending !== true)
+  }
+
+  /** Returns true if a pending parent should be cleared because all its direct children exist. */
+  const shouldClearPending = (val: MaybePending | null, index: Index<MaybePending | null>) => {
+    if (
+      val &&
+      val.pending === true &&
+      typeof val === 'object' &&
+      'childrenMap' in val &&
+      typeof (val as { childrenMap: unknown }).childrenMap === 'object'
+    ) {
+      const childIds = Object.values((val as { childrenMap: Record<string, string> }).childrenMap)
+      return childIds.length > 0 && childIds.every(cid => !!index[cid])
+    }
+    return false
+  }
+
+  /** Remove pending property from an object. */
+  const stripPending = (val: MaybePending): T => {
+    const { pending: _discard, ...rest } = val
+    return rest as T
+  }
+
+  /** True if object has a childrenMap and is pending. */
+  const isPending = (val: T | null) =>
+    !!(val as MaybePending)?.pending && typeof val === 'object' && 'childrenMap' in (val as object)
+
+  // Track parents that might be eligible for pending clearing
+  const maybePending = new Set<string>()
```

Comment:
Perhaps we could call this `pendingKeys` or `pendingThoughtKeys`. From my understanding, it's not that they are maybe pending, but they are actually pending and maybe should be cleared.

Reply:
Good idea. And based on this, I've refactored the line which I thought was still procedural. Kindly check.